pageserver/src/walingest.rs

```diff
@@ -104,12 +103,66 @@ struct WarnIngestLag {
     timestamp_invalid_msg_ratelimit: RateLimit,
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum WalIngestError {
+    #[error(transparent)]
+    #[allow(private_interfaces)]
+    PageReconstructError(#[from] PageReconstructError),
+    #[error(transparent)]
+    DeserializationFailure(#[from] DeserializeError),
+    #[error(transparent)]
+    SerializationFailure(#[from] SerializeError),
+    #[error("the request contains data not supported by pageserver: {0} @ {1}")]
+    InvalidKey(Key, Lsn),
+    #[error("twophase file for xid {0} already exists")]
+    FileAlreadyExists(u64),
+    #[error("slru segment {0:?}/{1} already exists")]
+    SlruAlreadyExists(SlruKind, u32),
+    #[error("relation already exists")]
+    RelationAlreadyExists,
+    #[error("invalid relnode")]
+    InvalidRelnode,
+    #[error("invalid reldir key")]
+    InvalidRelDirKey,
+
+    #[error(transparent)]
+    AssertionError(anyhow::Error),
+    #[error(transparent)]
+    EncodeAuxFileError(anyhow::Error),
+    #[error(transparent)]
+    MaybeRelSizeV2Error(anyhow::Error),
+    #[error(transparent)]
+    ToRelBlockErr(anyhow::Error),
+
+    #[error("timeline shutting down")]
+    Cancelled,
+}
+
+#[macro_export]
+macro_rules! ensure_walingest {
+    ($($t:tt)*) => {
+        _ = || -> Result<(), anyhow::Error> {
+            anyhow::ensure!($($t)*);
+            Ok(())
+        }().map_err(|e| WalIngestError::AssertionError(e));
+    };
+}
+
+#[macro_export]
+macro_rules! bail_walingest {
+    ($($t:tt)*) => {
+        _ = || -> Result<(), anyhow::Error> {
+            anyhow::bail!($($t)*);
+        }().map_err(|e| WalIngestError::AssertionError(e));
+    };
+}
```

Comment:
This is only used in one place. We could just explicitly return for net less code.

Reply:
removed (although now we'd be using it in two places)